repo-scripts/size-analysis/analysis-helper.ts

```diff
@@ -0,0 +1,316 @@
+    mangle: true,
+    compress: false
+  });
+
+  fs.unlinkSync(input);
+  fs.unlinkSync(output);
+  return { dependencies, sizeInBytes: Buffer.byteLength(code!, 'utf-8') };
+}
+
+/**
+ * Extracts all function, class and variable declarations using the TypeScript
+ * compiler API.
+ */
+export function extractDeclarations(
+  jsFile: string,
+  map?: Map<string, string>
+): MemberList {
+  const program = ts.createProgram([jsFile], { allowJs: true });
+  const checker = program.getTypeChecker();
+
+  const sourceFile = program.getSourceFile(jsFile);
+  if (!sourceFile) {
+    throw new Error('Failed to parse file: ' + jsFile);
+  }
+
+  let declarations: MemberList = {
+    functions: [],
+    classes: [],
+    variables: [],
+    enums: []
+  };
+
+  ts.forEachChild(sourceFile, node => {
+    if (ts.isFunctionDeclaration(node)) {
+      declarations.functions.push(node.name!.text);
+    } else if (ts.isClassDeclaration(node)) {
+      declarations.classes.push(node.name!.text);
+    } else if (ts.isVariableDeclaration(node)) {
+      declarations.variables.push(node.name!.getText());
+    } else if (ts.isEnumDeclaration(node)) {
+      declarations.enums.push(node.name.escapedText.toString());
+    } else if (ts.isVariableStatement(node)) {
+      const variableDeclarations = node.declarationList.declarations;
+      variableDeclarations.forEach(variableDeclaration => {
+        if (ts.isIdentifier(variableDeclaration.name)) {
+          declarations.variables.push(
+            (variableDeclaration.name as ts.Identifier).getText(sourceFile)
+          );
+        }
+        // TODO: variableDeclaration.name is an union type (Identifier | BindingPattern)
+        // need Identifier type, not sure in what case BindingPattern type is for.
+        else {
+          console.log(
+            'this VariableDeclaration.name object is of BindingPattern type !'
+          );
+        }
+      });
+    } else if (ts.isExportDeclaration(node)) {
+      if (node.moduleSpecifier && ts.isStringLiteral(node.moduleSpecifier)) {
+        const symbol = checker.getSymbolAtLocation(node.moduleSpecifier);
+        const reExportFullPath = symbol.valueDeclaration.getSourceFile()
+          .fileName;
+        const reExportsMember = extractDeclarations(reExportFullPath);
+
+        // named exports
+        if (node.exportClause && ts.isNamedExports(node.exportClause)) {
+          const actualExports: string[] = [];
+          node.exportClause.elements.forEach(exportSpecifier => {
+            const reExportedSymbol: string = extractRealSymbolName(
+              exportSpecifier
+            );
+            // if export is renamed, replace with new name
+            if (isExportRenamed(exportSpecifier)) {
+              actualExports.push(exportSpecifier.name.escapedText.toString());
+              replaceAll(
+                reExportsMember,
+                reExportedSymbol,
+                exportSpecifier.name.escapedText.toString()
+              );
```

Comment:
This part is confusing.  I guess `exportSpecifier.name.escapedText.toString()` is the renamed name, and `reExportedSymbol ` is the original name?

Can you add some comment to explain that, and also add comment to explain what `replaceAll()` tries to do.